consul/src/main/java/org/secondbase/consul/ConsulModule.java

```diff
@@ -0,0 +1,217 @@
+package org.secondbase.consul;
+
+import com.google.common.net.HostAndPort;
+import com.orbitz.consul.AgentClient;
+import com.orbitz.consul.Consul;
+import com.orbitz.consul.model.agent.Registration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.secondbase.core.SecondBase;
+import org.secondbase.core.SecondBaseException;
+import org.secondbase.core.config.SecondBaseModule;
+
+/**
+ * Consul module for SecondBase. Handles registration of services in Consul.
+ */
+public final class ConsulModule implements SecondBaseModule {
+
+    private static final Logger LOG = Logger.getLogger(ConsulModule.class.getName());
+
+    private static final ScheduledExecutorService consulKeepAlive
+            = Executors.newSingleThreadScheduledExecutor();
+    private static final long keepAliveInitialDelaySec = 1L;
+    private static final long keepAlivePeriodSec = 30L;
+    private static final DeregisterThread deregisterThread = new DeregisterThread();
+    private static final AtomicBoolean isDeregisterThreadActive = new AtomicBoolean(false);
+
+    private static ConsulModule consulModule;
+    private static AgentClient agentClient;
+
+    /**
+     * Deregister a list of services. To be used for deregistering consul services when
+     * shutting down.
+     */
+    private static class DeregisterThread extends Thread {
+        private Map<String, AgentClient> services = new HashMap<>();
+
+        /**
+         * Record a serviceid to be deregistered at shutdown
+         * @param serviceId the service id to be deregistered
+         * @param agentClient the consul client to use for deregistration
+         */
+        void add(final String serviceId, final AgentClient agentClient) {
+            services.put(serviceId, agentClient);
+        }
+
+        @Override
+        public void run() {
+            for (final Map.Entry<String, AgentClient> serviceReg : services.entrySet()) {
+                serviceReg.getValue().deregister(serviceReg.getKey());
+            }
+        }
+    }
+
+    @Override
+    public void load(final SecondBase secondBase) {
+        secondBase.getFlags().loadOpts(ConsulModuleConfiguration.class);
+    }
+
+    /**
+     * Used by SecondBase.
+     */
+    public ConsulModule() {}
+
+    // private constructor. use getInstance().
+    private ConsulModule(final String host) {
```

Comment:
If the consul agent object is injected here it gets very easy to write tests for this. So I would strongly recommend doing that.